// zeromq/lib/compat/long-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZMQ_MAX_SOCKETS = exports.ZMQ_IO_THREADS = exports.ZMQ_CONNECT_TIMEOUT = exports.ZMQ_HEARTBEAT_TIMEOUT = exports.ZMQ_HEARTBEAT_TTL = exports.ZMQ_HEARTBEAT_IVL = exports.ZMQ_ZAP_DOMAIN = exports.ZMQ_CURVE_SERVERKEY = exports.ZMQ_CURVE_SECRETKEY = exports.ZMQ_CURVE_PUBLICKEY = exports.ZMQ_CURVE_SERVER = exports.ZMQ_PLAIN_PASSWORD = exports.ZMQ_PLAIN_USERNAME = exports.ZMQ_PLAIN_SERVER = exports.ZMQ_MECHANISM = exports.ZMQ_IPV6 = exports.ZMQ_ROUTER_RAW = exports.ZMQ_XPUB_VERBOSE = exports.ZMQ_DELAY_ATTACH_ON_CONNECT = exports.ZMQ_TCP_ACCEPT_FILTER = exports.ZMQ_TCP_KEEPALIVE_INTVL = exports.ZMQ_TCP_KEEPALIVE_IDLE = exports.ZMQ_TCP_KEEPALIVE_CNT = exports.ZMQ_TCP_KEEPALIVE = exports.ZMQ_ROUTER_MANDATORY = exports.ZMQ_LAST_ENDPOINT = exports.ZMQ_IPV4ONLY = exports.ZMQ_SNDTIMEO = exports.ZMQ_RCVTIMEO = exports.ZMQ_MULTICAST_HOPS = exports.ZMQ_RCVHWM = exports.ZMQ_SNDHWM = exports.ZMQ_MAXMSGSIZE = exports.ZMQ_RECONNECT_IVL_MAX = exports.ZMQ_BACKLOG = exports.ZMQ_RECONNECT_IVL = exports.ZMQ_LINGER = exports.ZMQ_TYPE = exports.ZMQ_EVENTS = exports.ZMQ_FD = exports.ZMQ_RCVMORE = exports.ZMQ_RCVBUF = exports.ZMQ_SNDBUF = exports.ZMQ_RECOVERY_IVL_MSEC = exports.ZMQ_RECOVERY_IVL = exports.ZMQ_RATE = exports.ZMQ_UNSUBSCRIBE = exports.ZMQ_SUBSCRIBE = exports.ZMQ_IDENTITY = exports.ZMQ_AFFINITY = void 0;
exports.ZMQ_ROUTER_HANDOVER = void 0;
exports.ZMQ_AFFINITY = 4;
exports.ZMQ_IDENTITY = 5;
exports.ZMQ_SUBSCRIBE = 6;
exports.ZMQ_UNSUBSCRIBE = 7;
exports.ZMQ_RATE = 8;
exports.ZMQ_RECOVERY_IVL = 9;
exports.ZMQ_RECOVERY_IVL_MSEC = 9;
exports.ZMQ_SNDBUF = 11;
exports.ZMQ_RCVBUF = 12;
exports.ZMQ_RCVMORE = 13;
exports.ZMQ_FD = 14;
exports.ZMQ_EVENTS = 15;
exports.ZMQ_TYPE = 16;
exports.ZMQ_LINGER = 17;
exports.ZMQ_RECONNECT_IVL = 18;
exports.ZMQ_BACKLOG = 19;
exports.ZMQ_RECONNECT_IVL_MAX = 21;
exports.ZMQ_MAXMSGSIZE = 22;
exports.ZMQ_SNDHWM = 23;
exports.ZMQ_RCVHWM = 24;
exports.ZMQ_MULTICAST_HOPS = 25;
exports.ZMQ_RCVTIMEO = 27;
exports.ZMQ_SNDTIMEO = 28;
exports.ZMQ_IPV4ONLY = 31;
exports.ZMQ_LAST_ENDPOINT = 32;
exports.ZMQ_ROUTER_MANDATORY = 33;
exports.ZMQ_TCP_KEEPALIVE = 34;
exports.ZMQ_TCP_KEEPALIVE_CNT = 35;
exports.ZMQ_TCP_KEEPALIVE_IDLE = 36;
exports.ZMQ_TCP_KEEPALIVE_INTVL = 37;
exports.ZMQ_TCP_ACCEPT_FILTER = 38;
exports.ZMQ_DELAY_ATTACH_ON_CONNECT = 39;
exports.ZMQ_XPUB_VERBOSE = 40;
exports.ZMQ_ROUTER_RAW = 41;
exports.ZMQ_IPV6 = 42;
exports.ZMQ_MECHANISM = 43;
exports.ZMQ_PLAIN_SERVER = 44;
exports.ZMQ_PLAIN_USERNAME = 45;
exports.ZMQ_PLAIN_PASSWORD = 46;
exports.ZMQ_CURVE_SERVER = 47;
exports.ZMQ_CURVE_PUBLICKEY = 48;
exports.ZMQ_CURVE_SECRETKEY = 49;
exports.ZMQ_CURVE_SERVERKEY = 50;
exports.ZMQ_ZAP_DOMAIN = 55;
exports.ZMQ_HEARTBEAT_IVL = 75;
exports.ZMQ_HEARTBEAT_TTL = 76;
exports.ZMQ_HEARTBEAT_TIMEOUT = 77;
exports.ZMQ_CONNECT_TIMEOUT = 79;
exports.ZMQ_IO_THREADS = 1;
exports.ZMQ_MAX_SOCKETS = 2;
exports.ZMQ_ROUTER_HANDOVER = 56;
